scripts/generateLocalizedDiagnosticMessages.ts

```diff
@@ -0,0 +1,139 @@
+import * as fs from "fs";
+import * as path from "path";
+import * as xml2js from "xml2js";
+
+function main(): void {
+    const args = process.argv.slice(2);
+    if (args.length != 3) {
+        console.log("Usage:")
+        console.log("\tnode generateLocalizedDiagnosticMessages.js <lcl source directory> <output directory> <generated diagnostics map file>");
+        return;
+    }
+
+    const inputPath = args[0];
+    const outputPath = args[1];
+    const diagnosticsMapFilePath = args[2];
+
+    // generate the lcg file for enu
+    generateLCGFile();
+
+    // generate other langs
+    fs.readdir(inputPath, (err, files) => {
+        handelError(err);
+        files.forEach(visitDirectory);
+    });
+
+    return;
+
+    function visitDirectory(name: string) {
+        const inputFilePath = path.join(inputPath, name, "diagnosticMessages", "diagnosticMessages.generated.json.lcl");
+        const outputFilePath = path.join(outputPath, name, "diagnosticMessages.generated.json");
+        fs.readFile(inputFilePath, (err, data) => {
+            handelError(err);
+            xml2js.parseString(data.toString(), (err, result) => {
+                handelError(err);
+                writeFile(outputFilePath, xmlObjectToString(result));
+            });
+        });
+    }
+
+    function handelError(err: null | object) {
+        if (err) {
+            console.error(err);
+            process.exit(1);
+        }
+    }
+
+    function xmlObjectToString(o: any) {
+        const out: any = {};
+        for (const item of o["LCX"]["Item"][0]["Item"][0]["Item"]) {
+            let ItemId = item["$"]["ItemId"];
+            let Val = item["Str"][0]["Tgt"] ? item["Str"][0]["Tgt"][0]["Val"][0] : item["Str"][0]["Val"][0];
+
+            if (typeof ItemId !== "string" || typeof Val !== "string") {
+                console.error("Unexpected XML file structure");
+                process.exit(1);
+            }
+
+            if (ItemId.charAt(0) === ";") {
+                ItemId = ItemId.slice(1); // remove leading semicolon
+            }
+
+            Val = Val.replace(/]5D;/, "]"); // unescape `]`
+            out[ItemId] = Val;
+        }
+        return JSON.stringify(out, undefined, 2);
+    }
+
+
+    function ensureDirectoryExists(directoryPath: string, action: () => void) {
+        fs.exists(directoryPath, exists => {
+            if (!exists) {
+                const basePath = path.dirname(directoryPath);
+                if (basePath !== directoryPath) {
+                    return ensureDirectoryExists(basePath, () => fs.mkdir(directoryPath, action));
+                }
+            }
+            action();
+        });
+    }
+
+    function writeFile(fileName: string, contents: string) {
+        ensureDirectoryExists(path.dirname(fileName), () => {
+            fs.writeFile(fileName, contents, handelError);
+        });
+    }
+
+    function objectToList(o:  Record<string, string>) {

```

Comment:
If we're assuming an ES2015 runtime, you can use `Object.entries` instead, but this is probably fine.

Reply:
We already use `const` in some of our build scripts, I believe, and haven't received any complaints. We only build on back to Node 4, so it seems reasonable.

Reply:
`Object.entries` is ES2017. not sure if was on node v4 or not..

Reply:
According to http://node.green/, node `7.5` was the first with `Object.entries`... so probably not yet. However `map.entries` has been available since node 4. 🐱